cilium/cmd/preflight_k8s_valid_cnp.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/cilium/cilium/pkg/k8s"
+	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/scheme"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/option"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	apiextensionsinternal "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/validation"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+var validateCNP = &cobra.Command{
+	Use:   "validate-cnp",
+	Short: "Validate Cilium Network Policies deployed in the cluster",
+	Long: `Before upgrading Cilium it is recommended to run this validation checker
+to make sure the policies deployed are valid. The validator will verify if all policies
+deployed in the cluster are valid, in case they are not, an error is printed and the
+has an exit code -1 is returned.`,
+	Run: func(cmd *cobra.Command, args []string) {
+		err := validateCNPs()
+		if err != nil {
+			log.Error(err)
+			os.Exit(-1)
+		}
+	},
+}
+
+const validateK8sPoliciesTimeout = 5 * time.Minute
+
+func validateCNPs() error {
+	// The internal packages log things. Make sure they follow the setup of of
+	// the CLI tool.
+	logging.DefaultLogger.SetFormatter(log.Formatter)
+
+	log.Info("Setting up Kubernetes client")
+
+	k8sClientQPSLimit := viper.GetFloat64(option.K8sClientQPSLimit)
+	k8sClientBurst := viper.GetInt(option.K8sClientBurst)
+
+	k8s.Configure(k8sAPIServer, k8sKubeConfigPath, float32(k8sClientQPSLimit), k8sClientBurst)
+
+	if err := k8s.Init(); err != nil {
+		log.WithError(err).Fatal("Unable to connect to Kubernetes apiserver")
+	}
+
+	ctx, initCancel := context.WithTimeout(context.Background(), validateK8sPoliciesTimeout)
```

Comment:
Should we have a real timeout here, and for the other contexts around this area?

Reply:
I don't think so because this is a user request and we don't know the number of CNPs installed to have a timeout here. Also, this will be triggered by the user so the user can cancel the request with CTRL+C

Reply:
I'm thinking from the perspective of this being run automatically as part of the preflight. What if some issue causes APIserver connectivity breakage then, will the pod just hang in not-ready state? How will users know?